### PR TITLE
TypeScript and Custom Iterable

### PR DESCRIPTION
**Pull request by Zalastax**
I needed types for this library for another project and want to give you the opportunity to merge my changes which adds type definitions. The changes also include a rewrite of IntervalTree.inOrder and IntervalTree.preOrder to remove the dependency on regeneratorRuntime. This leads to leaner and faster generated code for inOrder and preOrder with the disadvantage of a somewhat less elegant implementation. The only changes to the API are that inOrder and preOrder return an iterable of records instead of an iterable of the records data. My motivation for the API change is that consumers easily can extract the data property from the record and that it previously was impossible to retrieve the interval associated with the data.

### CODE REVIEW

**Comment by 2Pacalypse-** [comment]
Thanks a lot for the PR!

I'm not really familiar with TypeScript so it's gonna take me a bit to review the changes. Though from what I can tell, it looks like using TypeScript doesn't affect the end user at all?

Also, I'm not really sure about the API change. The reason why I didn't return the Interval itself when searching is because that class is an implementation detail, and the user should operate with `data` objects. You can put interval into the data if you really need it. Although, I guess I can see some use cases where you only care about the intervals without having any data associated with them, so it's a little bit inconvenient to work with the library in that case.

PS: I agree with nikteg, that extra space is a bit jarring :)

**Comment by Zalastax** [comment]
I'll fix the spacing and rebase the commits before merging. Anyone using JavaScript won't see a difference when using the package (the typescript compiler transpiles it to ES5 just like babel). So the change of language only affects maintainers and typescript users.

I acknowledge your viewpoint that records are implementation details but as a library user I expect the intervals to be available without relying on storing extra data. It only complicates implementation and usage. Records are very uncomplicated objects so I see no harm in exposing them.

I have another proposition. Instead of taking intervals as two separate numbers, they could be received as one object with a low and high field. Search and traversal would then return these user created objects. The user will be able to add the fields they need as they control the object. This solves the problem of exposing implementation while simplifying the API. I'll make a commit in a little while to show you what I mean.

**Comment by 2Pacalypse-** [comment]
Gotta say, I'm not really a fan of this API change. Not sure how it simplifies the API, when it replaces exposing one implementation detail from your previous version with another. Also requires users to do more work than before (note: this is similar to how the API first worked before it was refactored).

Another thing is that TypeScript shouldn't be the main use case of the lib, so demointervaltree.js should show how to use the lib without TypeScript and a new file to demo TypeScript usage.

The main thing I like about the current API is that when the user sees the example in the README, the simplicity of using it is really obvious (I hope ^^). Granted, searching for intervals that don't have any data associated with it is a little bit unintuitive, so I wouldn't be adverse to changing that.

In any case, the changes here are substantial (not to API alone), so I can't even think of merging it before I'm certain that I can continue working on it or fix any bugs, if needed. That requires me learning TypeScript, which I actually just started to do few days ago!

**Comment by Zalastax** [comment]
I see your point about the API and will add a wrapper class so it's possible to use it the old way by default and with the other api by another export. I think the way I propose is conceptually simpler: insert any object with properties "low" and "high" and they will be returned by search or iteration, just like with other collections.

Rewerting to ES2015 is really simple if you don't like TypeScript; Strip the types by hand or use target es6 as compiler setting. You shouldn't have trouble adapting, as you already know modern JavaScript.

I hope you decide to merge eventually but I understand if you decide not to. I needed to make the changes anyway so no biggie.

**Comment by 2Pacalypse-** [comment]
I did a quick test and it seems like the search is broken? The following code returns "No overlapping intervals":

```
import IntervalTree from 'node-interval-tree'
import cuid from 'cuid'

const intervalTree = new IntervalTree()

function getRandomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1)) + min
}

for (let i = 1; i <= 100; i++) {
  let low = getRandomInt(0, 100)
  let high = getRandomInt(0, 100)

  if (high < low) {
    const temp = high
    high = low
    low = temp
  }

  intervalTree.insert({low, high, data: cuid()})
}

console.log('Number of the records in the tree: ' + intervalTree.count)

const results = intervalTree.search(0, 100)
if (!results || results.length === 0) {
  console.log('No overlapping intervals')
} else {
  console.log('Found ' + results.length + ' overlapping intervals')

  for (let i = 0; i < results.length; i++) {
    console.log(results[i])
  }
}
```

**Comment by Zalastax** [comment]
Please add it as a test and push the commit to my repo (you should have access).  I'll take a look in a few hours.

**Comment by 2Pacalypse-** [comment]
Oops, not sure if you wanted me to write an actual test case for this, but I'm at work so I don't really have time to do that.

Pushed the file itself which I now realize is pretty useless, as you could've copy-pasted it as well,, but whatever ^^

**Comment by Zalastax** [comment]
The import was not using a relative path so it must have imported and old version that you had installed or linked in your global module directory. It seems to work fine after changing the import.

**Comment by 2Pacalypse-** [comment]
Ah, that was probably it. Good work ^^

**Comment by Zalastax** [comment]
I created a wrapper for IntervalTree that uses the 0.2.1 interface. I'm quite happy with this solution as it exposes the simple interface first, without limiting the advances user.

**Comment by 2Pacalypse-** [comment]
Looks good. I'll probably merge it, eventually, once I get some time to actually take a closer look at all the changes.

**Comment by Zalastax** [comment]
Great! I'll update the README according to the changes. Drop a comment here when you're ready so I can clean the commits up. Do you want everything to be one commit for the whole pull request or multiple granular?

**Comment by 2Pacalypse-** [comment]
I think multiple commits makes more sense and will make it easier to review the changes.

**Comment by Zalastax** [comment]
@2Pacalypse- I've cleaned things up and would love to hear your feedback!

**Comment by 2Pacalypse-** [comment]
Looks great, I'll take a deeper look over the weekend and hopefully get it merged ^^.

**Comment by 2Pacalypse-** [comment]
Hmm, is `data` not an optional parameter anymore?

**Comment by Zalastax** [comment]
The "easy to use" implementation can technically be used without the data parameter. In TypeScript you'd use `IntervalTree<undefined>` and `.insert(low, high, undefined)` while in JavaScript it can be used just like before. I'd recommend using the advanced export instead so that .search doesn't return an array full of undefined.

**Comment by 2Pacalypse-** [comment]
Yeah, I was thinking we should indicate that `data` param is optional in the README for the basic use case, but it really doesn't make much sense to advocate that, since search is broken for all intents and purposes in that case.

Anyway, this looks good to be merged, so I'm gonna do that once I come home from the work. Once again, thanks a lot for your work, and sorry it took so long to merge this; I'm using this library in one of the private projects I'm working on and I wanted to make sure it will stay unaffected by this change.

File: demo/dataintervaltree.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import DataIntervalTree from '../index' // import IntervalTree from 'node-interval-tree'
+import  cuid = require('cuid')
```
Comment (2Pacalypse-) [inline demo/dataintervaltree.ts]: Nit: extra space between `import` and `cuid`

Comment (2Pacalypse-) [reply demo/dataintervaltree.ts]: Also, any particular reason this is using `require` instead of `from`?

File: demo/intervaltree.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+import { IntervalTree } from '../lib/index.js' // import IntervalTree from 'node-interval-tree'
+const cuid = require('cuid')
```
Comment (2Pacalypse-) [inline demo/intervaltree.js:2]: `from` instead of `require`?

File: demo/intervaltree.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
+import { IntervalTree, Interval } from '../index' // import IntervalTree from 'node-interval-tree'
+import cuid = require('cuid')
```
Comment (2Pacalypse-) [inline demo/intervaltree.ts:2]: `from` instead of `require`

File: index.ts [hunk 4]
```diff
@@ -4,87 +4,72 @@
 // Both insertion and deletion require O(log n) time. Searching requires O(k*logn) time, where `k`
 // is the number of intervals in the output list.
 
-import is from 'immutable-is'
+import isSame = require('shallowequal')
```
Comment (2Pacalypse-) [inline index.ts:7]: `from` instead of `require`?

File: test/test.ts [hunk 5]
```diff
@@ -0,0 +1,275 @@
...
+      expect(searchResult[0].data).to.eql('data1')
+      expect(searchResult[1].data).to.eql('data2')
+    })
+
+    it('should reject intervals where low > high', () => {
+      const tree = new IntervalTree<Interval>()
+      const high = 10
+      const low = 15
+
+
```
Comment (2Pacalypse-) [inline test/test.ts]: Nit: extra new line :d

File: index.ts [hunk 6]
```diff
@@ -4,87 +4,72 @@
...
+  public max: number
+  public records: T[] = []
+  public parent?: Node<T>
+  public height = 0
+  public left?: Node<T>
+  public right?: Node<T>
 
-    // Save the array of all records with the same key for this node
-    this.records = []
-    this.records.push(record)
```
Comment (2Pacalypse-) [inline index.ts]: Nit: extra new line

File: index.ts [hunk 7]
```diff
@@ -252,46 +241,48 @@ class Node {
     leftChild.updateHeight()
   }
 
-  // Rebalances the tree if the height value between two nodes of the same parent is greater than
+   // Rebalances the tree if the height value between two nodes of the same parent is greater than
```
Comment (2Pacalypse-) [inline index.ts]: Nit: extra space

**Comment by Zalastax** [comment]
Great! The important thing is that you are comfortable. I had no problem using my fork in the meantime. I'll go over your comments today. The requires might be because of how the declarations are defined.

**Comment by Zalastax** [comment]
@2Pacalypse- Cleanup done. Whitespace is my greatest enemy! Replace typings with npm @types
Tried changing 'cuid' to import but it can't be done it "resolves to a non-entity module and cannot be imported using this construct"

**Comment by 2Pacalypse-** [comment]
Sorry it took so long, again, but finally managed to test this on a project where I'm using it and everything seems to be working fine, so it is now merged :)

**Comment by Zalastax** [comment]
Great! The released version is working for me as well!

**Pull request merged**